src/conways.rs

```diff
@@ -1,25 +1,263 @@
+
+        //check that center cell has 2 neighbors
+        assert_eq!(grid.count_neighbors(1, 1), 2);
+
+        //check that center cell lives after next generation
+        grid.next_cell_generation();
+        assert_eq!(grid.get(1, 1), CellState::Alive);
+    }
+
+    // Rule 3: Any dead cell with exactly three live neighbours comes to life.
+    #[test]
+    fn test_rule_3() {
+        let mut grid = Grid::new(3, 3);
+        grid.set_alive(0, 0); // Set top-left cell Alive
+        grid.set_alive(0, 1); // Set top-middle cell Alive
+        grid.set_alive(1, 0); // Set middle-left cell Alive
+
+        //check that center cell has 3 neighbors
+        assert_eq!(grid.count_neighbors(1, 1), 3);
+
+        //check that center cell comes to life after next generation
+        println!("Before:");
+        print_grid(&grid);
+
+        grid.next_cell_generation();
+
+        println!("After:");
+        print_grid(&grid);
+
+        assert_eq!(grid.get(1, 1), CellState::Alive);
+    }
+
+    // Rule 4: Any live cell with more than three live neighbours dies (referred to as overpopulation).
+    #[test]
+    fn test_rule_4() {
+        let mut grid = Grid::new(3, 3);
+        grid.set_alive(1, 1); // Set center cell Alive
+        grid.set_alive(0, 0); // Set top-left cell Alive
+        grid.set_alive(0, 1); // Set top-middle cell Alive
+        grid.set_alive(0, 2); // Set top-right cell Alive
+        grid.set_alive(1, 0); // Set middle-left cell Alive
+
+        //check that center cell has 4 neighbors
+        assert_eq!(grid.count_neighbors(1, 1), 4);
+
+        //check that center cell dies after next generation
+        grid.next_cell_generation();
+        assert_eq!(grid.get(1, 1), CellState::Dead);
+    }
+
+    fn print_grid(grid: &Grid) {
+        for row in &grid.grid {
+            for cell in row {
+                match cell {
+                    CellState::Dead => print!(". "),
+                    CellState::Alive => print!("O "),
+                }
+            }
+            println!();
+        }
+        println!();
+    }
+
+    #[test]
+    fn test_blinker_pattern() {
+        let mut grid = Grid::new(5, 5);
+
+        grid.set_alive(2, 1);
+        grid.set_alive(2, 2);
+        grid.set_alive(2, 3);
+
+        println!("Initial generation:");
+        print_grid(&grid);
+
+        grid.next_cell_generation();
+        println!("Next generation:");
+        print_grid(&grid);
+    }
+
+    #[test]
+    fn test_toad_pattern() {
+        let mut grid = Grid::new(6, 6);
+
+        // Set toad pattern
+        grid.set_alive(2, 2);
+        grid.set_alive(3, 2);
+        grid.set_alive(4, 2);
+        grid.set_alive(1, 3);
+        grid.set_alive(2, 3);
+        grid.set_alive(3, 3);
+
+        println!("Initial generation:");
+        print_grid(&grid);
+
+        grid.next_cell_generation();
+        println!("Next generation:");
+        print_grid(&grid);
+    }
 }
```

Comment:
This file is missing a blank line at the end; you should configure your editor to add it automatically.

Reply:
Will do it in next PR